### PR TITLE
Experimental: Check github for update on launch once a day (set checkRssUpdate=true)

### PR DESCRIPTION
**Pull request by Rinn**
Queries the GitHub api for the current release and prompts if a newer release is available. Basically the same style update check as the sourceforge point release update check, but behind a setting now. Only prompts to update on launch once a day.

Possible improvements for additional PRs
- Add checkRssUpdate option to preferences window
- Add manual update check that bypasses settings \ daily limit
- Query and show change log in app
- Download jar directly \ auto-relaunch?
- Disable update check if mafia is built locally

### CODE REVIEW

File: src/net/sourceforge/kolmafia/KoLmafia.java [hunk 1]
```diff
@@ -660,7 +670,8 @@ public static void refreshSession() {
 
     boolean shouldResetCounters = false;
     boolean shouldResetGlobalCounters = false;
-    // Assume if rollover has changed by an hour, it is a new rollover. Time varies slightly between
+    // Assume if rollover has changed by an hour, it is a new rollover. Time varies
+    // slightly between
     // servers by a few seconds.
```
Comment (heeheehee-kolmafia) [inline src/net/sourceforge/kolmafia/KoLmafia.java:675]: Can you reflow this comment? Now you have two short lines next to each other.

ditto below. Do you have your column limit set to 80 chars? 100 is the standard in Google Java Style.

File: src/net/sourceforge/kolmafia/KoLmafia.java [hunk 2]
```diff
@@ -1861,15 +1874,79 @@ public static String whoisPlayer(final String player) {
...
+      if (now - lastUpdate < 86400000L) {
+        return;
+      }
+
+      String contents = null;
+
+      try {
+        BufferedReader reader =
+            FileUtilities.getReader(
+                "https://api.github.com/repos/kolmafia/kolmafia/releases/latest");
```
Comment (heeheehee-kolmafia) [inline src/net/sourceforge/kolmafia/KoLmafia.java:1901]: Consider reframing this as a try-with-resources statement.

This has the potential to leak resources if it throws an exception before getting to reader.close().

File: src/net/sourceforge/kolmafia/KoLmafia.java [hunk 3]
```diff
@@ -1861,15 +1874,79 @@ public static String whoisPlayer(final String player) {
...
+      Preferences.setString("lastRssVersion", currentVersion);
+      Preferences.setString("lastRssUpdate", Long.toString(now));
+
+      if (currentVersionInt <= StaticEntity.getRevision() || currentVersion.equals(lastVersion)) {
+        return;
+      }
+
+      if (InputFieldUtilities.confirm(
+          String.format(
+              "A new version KoLmafia r%s is now available. Would you like to download it now?",
```
Comment (heeheehee-kolmafia) [inline src/net/sourceforge/kolmafia/KoLmafia.java:1946]: formatting nit: perhaps "new version of KoLmafia (r%s) is now available"?  This reads a bit awkwardly to me as it is right now.

(Maybe that's just me)

File: src/data/defaults.txt [hunk 4]
```diff
@@ -17,6 +17,7 @@ global	chatBeep	false
 global	chatFontSize
 global	chatLinksUseRelay	false
 global	chatStyle	0
+global	checkRssUpdate	false
```
Comment (heeheehee-kolmafia) [inline src/data/defaults.txt:20]: Do you also need to set a default value for lastRssUpdate?

I'm worried that Long.parseLong("") will throw an exception.

File: src/net/sourceforge/kolmafia/KoLmafia.java [hunk 5]
```diff
@@ -1861,15 +1874,79 @@ public static String whoisPlayer(final String player) {
...
   private static class UpdateCheckRunnable implements Runnable {
     public void run() {
-      // TODO: Check for new version on jenkins\github after migration is complete. See revision
-      // history for old release update check.
+      boolean checkUpdate = Preferences.getBoolean("checkRssUpdate");
+      if (!checkUpdate) {
+        return;
+      }
+
+      long lastUpdate = Long.parseLong(Preferences.getString("lastRssUpdate"));
```
Comment (heeheehee-kolmafia) [inline src/net/sourceforge/kolmafia/KoLmafia.java:1890]: Why not just Preferences.getLong?

File: src/net/sourceforge/kolmafia/KoLmafia.java [hunk 6]
```diff
@@ -1861,15 +1874,79 @@ public static String whoisPlayer(final String player) {
...
+      try {
+        currentVersionInt = Integer.parseInt(currentVersion) + 1;
+      } catch (NumberFormatException e) {
+        System.out.println(e.getMessage() + " while parsing version number");
+        return;
+      }
+
+      String lastVersion = Preferences.getString("lastRssVersion");
+      Preferences.setString("lastRssVersion", currentVersion);
+      Preferences.setString("lastRssUpdate", Long.toString(now));
```
Comment (heeheehee-kolmafia) [inline src/net/sourceforge/kolmafia/KoLmafia.java:1938]: similarly, Preferences.setLong?

File: src/net/sourceforge/kolmafia/KoLmafia.java [hunk 7]
```diff
@@ -1861,15 +1874,79 @@ public static String whoisPlayer(final String player) {
...
+
+      if (currentVersion == null) {
+        return;
+      }
+
+      int currentVersionInt;
+      try {
+        currentVersionInt = Integer.parseInt(currentVersion) + 1;
+      } catch (NumberFormatException e) {
+        System.out.println(e.getMessage() + " while parsing version number");
```
Comment (heeheehee-kolmafia) [inline src/net/sourceforge/kolmafia/KoLmafia.java:1932]: Perhaps also print the current version in the error message?

e.g. System.out.println(e.getMessage() + " while parsing version number ' " + currentVersion + " ' ");

(extraneous spaces added only to make the single-quotes easier to see)

**Comment by BadHorseMonkey** [comment]
If this wasn’t small and a return to pre-watershed parity, I’d say “just implement update4J”. 

But it isn’t a lot of code and it may even lead the way to that or a similar tool.

File: src/net/sourceforge/kolmafia/KoLmafia.java (hunk 1 above)
Comment (Rinn) [reply src/net/sourceforge/kolmafia/KoLmafia.java:675]: oh i think this was an auto format in my editor, i'll revert them

File: src/net/sourceforge/kolmafia/KoLmafia.java (hunk 5 above)
Comment (Rinn) [reply src/net/sourceforge/kolmafia/KoLmafia.java:1890]: this is an existing preference, I didn't want to change it from a string

File: src/net/sourceforge/kolmafia/KoLmafia.java (hunk 6 above)
Comment (Rinn) [reply src/net/sourceforge/kolmafia/KoLmafia.java:1938]: existing preference, didn't want to change the type

File: src/data/defaults.txt (hunk 4 above)
Comment (Rinn) [reply src/data/defaults.txt:20]: It already does, it's an existing preference and it defaults to 0

**Comment by Rinn** [comment]
> If this wasn’t small and a return to pre-watershed parity, I’d say “just implement update4J”.
> 
> But it isn’t a lot of code and it may even lead the way to that or a similar tool.

Oh okay, let's just do that then.

**Pull request closed without merging**